src/Features/Core/Portable/Completion/FileSystemCompletionHelper.cs

```diff
@@ -0,0 +1,271 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.IO;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Shared.Utilities;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Completion
+{
+    internal class FileSystemCompletionHelper
+    {
+        private static char[] s_windowsDirectorySeparator = { '\\' };
+
+        private readonly Glyph _folderGlyph;
+        private readonly Glyph _fileGlyph;
+
+        // absolute paths
+        private readonly ImmutableArray<string> _searchPaths;
+        private readonly string _baseDirectoryOpt;
+
+        private readonly ImmutableArray<string> _allowableExtensions;
+
+        private readonly Lazy<string[]> _lazyGetDrives;
+        private readonly CompletionItemRules _itemRules;
+
+        public FileSystemCompletionHelper(
+            Glyph folderGlyph,
+            Glyph fileGlyph,
+            ImmutableArray<string> searchPaths,
+            string baseDirectoryOpt,
+            ImmutableArray<string> allowableExtensions,
+            CompletionItemRules itemRules)
+        {
+            Debug.Assert(searchPaths.All(path => PathUtilities.IsAbsolute(path)));
+            Debug.Assert(baseDirectoryOpt == null || PathUtilities.IsAbsolute(baseDirectoryOpt));
+
+            _searchPaths = searchPaths;
+            _baseDirectoryOpt = baseDirectoryOpt;
+            _allowableExtensions = allowableExtensions;
+            _folderGlyph = folderGlyph;
+            _fileGlyph = fileGlyph;
+            _itemRules = itemRules;
+
+            _lazyGetDrives = new Lazy<string[]>(GetLogicalDrives);
+        }
+
+        // virtual for testing
+        protected virtual string[] GetLogicalDrives() => 
+            IOUtilities.PerformIO(CorLightup.Desktop.GetLogicalDrives, Array.Empty<string>());
+
+        // virtual for testing
+        protected virtual bool DirectoryExists(string fullPath)
+        {
+            Debug.Assert(PathUtilities.IsAbsolute(fullPath));
+            return Directory.Exists(fullPath);
+        }
+
+        // virtual for testing
+        protected virtual IEnumerable<string> EnumerateDirectories(string fullDirectoryPath)
+        {
+            Debug.Assert(PathUtilities.IsAbsolute(fullDirectoryPath));
+            try
+            {
+                return Directory.EnumerateDirectories(fullDirectoryPath);
+            }
+            catch
```

Comment:
We have an IOUtilities class that will catch the appropriate exceptions so that we don't need catch-alls in our code when doing IO.